### DISCUSSION TITLE
The need for .XYZ files

### DISCUSSION THREAD

**Discussion by antbern** [karttapullautin/karttapullautin#121] [Ideas]
Currently one part of the execution that takes quite some time (at least for me) is the extraction of the binary `las/laz` files into the text-based (and in the case of from the compressed laz files, 10x bigger in size) `xyz` files. These then need to be read and parsed every time they are processed which for me seems like unnecessary work if we can just iterate the binary files directly :rocket: Reading and writiting to files is generally much slower than accessing data in RAM, for example, so we could also consider reading the points into memory once and reusing them there. But that is a memory-performance tradeoff.

So before I start looking at the code, is there any specific use of the XYZ files? I see there is some "thinning" going in while converting also, but is this necessary? (If the file is already a XYZ file it seems it skips the thinning). If thinning is needed we could at least store the points in a binary format, or implement that in another way (I have some ideas).

**Comment by rphlo** [comment]
I have thought the same...
The xyz file is obviously not the best... We could instead of use laz/las files as the las-rs library also allow writes... it would also allow faster boundaries checks thanks to the values being in the headers.
The need for intermediate file is there for some thinning and other filtering...
Having those in memory is not where i would go as some of those file can go upwards of 2GB so the in batch mode we might have difficulties...

**Reply by antbern** [reply]
Yes keeping the 2gb in memory is definitely not a good idea... perhaps it could be a user option to use if the file is just a few 100 MB (and there is a lot of ram available on the system) and then one can opt-in to better performance at the cost of memory usage 🤔 

Nevertheless it seems that a good next step might be to replace the XYZ file with Las/laz for now 👍

**Comment by rphlo** [comment]
I created a test branch https://github.com/rphlo/karttapullautin/tree/binary-files for the use of binary files instead of text XYZ files... (only batch mode might work)
So far the performance takes a hit... 💩
Maybe @antbern can read through my changes and spot some optimisations errors...

**Reply by antbern** [reply]
Nice, I'll take a look! This week is a bit packet but next week I have more time. 

From a quick glance though, I would say skip all the `unsafe` and C-like structure casts / transmutations and just use the interface in the `las` crate for writing and reading las files instead. Then you don't have to roll your own binary format and it is also an openly standardized format 🚀

**Reply by rphlo** [reply]
the las writing proved to be difficult with current library...

**Reply by antbern** [reply]
> the las writing proved to be difficult with current library.

Hmm, from the documentation it looked quite simple but perhaps it does not fit well into the current setup 🤔 

I can also have a look whenever I get time.
